File: samples/java/spark/sources/SparkKmeansDense.java
/* file: SparkKmeansDense.java */
/*******************************************************************************
* Copyright 2017-2019 Intel Corporation
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*******************************************************************************/

/*
//  Content:
//      Java sample of K-Means clustering in the distributed processing mode
////////////////////////////////////////////////////////////////////////////////
*/

package DAAL;

import java.util.ArrayList;
import java.util.List;
import java.util.Arrays;
import java.util.Map;
import java.util.HashMap;

import org.apache.spark.api.java.*;
import org.apache.spark.api.java.function.*;
import org.apache.spark.SparkConf;

import scala.Tuple2;
import com.intel.daal.algorithms.kmeans.*;
import com.intel.daal.algorithms.kmeans.init.*;
import com.intel.daal.data_management.data.*;
import com.intel.daal.services.*;

import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;

public class SparkKmeansDense {
    /* Class containing the algorithm results */
    static class KmeansResult {
        public HomogenNumericTable centroids;
    }
    private static final long nBlocks        = 4;
    private static final long nClusters      = 20;
    private static final int nIterations     = 5;
    private static final int nVectorsInBlock = 2500;

    public static KmeansResult runKmeans(DaalContext context, JavaPairRDD<Integer, HomogenNumericTable> dataRDD) {

        JavaRDD<InitPartialResult> partsRDD = computeInitLocal(context, dataRDD);
        HomogenNumericTable centroids       = computeInitMaster(context, partsRDD);

        PartialResult partRes;
        JavaRDD<PartialResult> partResRDD;
        for(int it = 0; it < nIterations; it++) {
            centroids.pack();
            partResRDD = computeLocal(context, dataRDD, centroids);
            centroids  = computeMaster(context, partResRDD);
        }

        KmeansResult result = new KmeansResult();
        result.centroids = centroids;
        return result;
    }

    private static JavaRDD<InitPartialResult> computeInitLocal(DaalContext context, JavaPairRDD<Integer, HomogenNumericTable> dataRDD) {
        return dataRDD.map(new Function<Tuple2<Integer, HomogenNumericTable>, InitPartialResult>() {
            public InitPartialResult call(Tuple2<Integer, HomogenNumericTable> tup) {

                DaalContext localContext = new DaalContext();

                /* Create an algorithm to initialize the K-Means algorithm on local nodes */
                InitDistributedStep1Local kmeansLocalInit = new InitDistributedStep1Local(localContext, Double.class, InitMethod.randomDense,
                                                                                          nClusters, nBlocks * nVectorsInBlock,
                                                                                          nVectorsInBlock * tup._1);
                /* Set the input data on local nodes */
                tup._2.unpack(localContext);
                kmeansLocalInit.input.set(InitInputId.data, tup._2);

                /* Initialize the K-Means algorithm on local nodes */
                InitPartialResult pres = kmeansLocalInit.compute();
                pres.pack();
                tup._2.pack();

                localContext.dispose();
                return pres;
            }
        });
    }

    private static HomogenNumericTable computeInitMaster(DaalContext context, JavaRDD<InitPartialResult> partsRDD) {
        /* Create an algorithm to compute k-means on the master node */
        InitDistributedStep2Master kmeansMasterInit = new InitDistributedStep2Master(context, Double.class,
                                                                                     InitMethod.randomDense, nClusters);
        /* Set the partial results recieved from the local nodes */
        List<InitPartialResult> partResList = partsRDD.collect();
        for (InitPartialResult value : partResList) {
            if(value != null) {
                value.unpack(context);
                kmeansMasterInit.input.add(InitDistributedStep2MasterInputId.partialResults, value);
            }
        }
        /* Compute initial centroids on the master node */
        kmeansMasterInit.compute();

        /* Finalize computations and retrieve the results */
        InitResult initResult = kmeansMasterInit.finalizeCompute();

        return (HomogenNumericTable)initResult.get(InitResultId.centroids);
    }

    private static JavaRDD<PartialResult> computeLocal(DaalContext context,
                                                       JavaPairRDD<Integer, HomogenNumericTable> dataRDD,
                                                       final NumericTable centroids) {
        return dataRDD.map(new Function<Tuple2<Integer, HomogenNumericTable>, PartialResult>() {
            public PartialResult call(Tuple2<Integer, HomogenNumericTable> tup) {
                DaalContext localContext = new DaalContext();

                /* Create an algorithm to compute k-means on local nodes */
                DistributedStep1Local kmeansLocal = new DistributedStep1Local(localContext, Double.class, Method.defaultDense, nClusters);

                tup._2.unpack(localContext);
                centroids.unpack(localContext);

                /* Set the input data on local nodes */
                kmeansLocal.input.set(InputId.data, tup._2);
                kmeansLocal.input.set(InputId.inputCentroids, centroids);

                /* Compute k-means on local nodes */
                PartialResult pres = kmeansLocal.compute();

                pres.pack();
                tup._2.pack();
                centroids.pack();

                localContext.dispose();

                return pres;
            }
        });
    }

    private static HomogenNumericTable computeMaster(DaalContext context, JavaRDD<PartialResult> partResRDD) {
        /* Create an algorithm to compute k-means on the master node */
        DistributedStep2Master kmeansMaster = new DistributedStep2Master(context, Double.class, Method.defaultDense, nClusters);

        /* Set the partial result to the master algorithm to compute the final result */
        List<PartialResult> partResList = partResRDD.collect();
        for (PartialResult value : partResList) {
            if(value != null) {
                value.unpack(context);
                kmeansMaster.input.add(DistributedStep2MasterInputId.partialResults, value);
            }
        }

        /* Compute k-means on the master node */
        kmeansMaster.compute();

        /* Finalize computations and retrieve the results */
        Result res = kmeansMaster.finalizeCompute();
        return (HomogenNumericTable)res.get(ResultId.centroids);
    }
}

File: samples/python/spark/sources/spark_KmeansDense.py
# file: spark_KmeansDense.py
#===============================================================================
# Copyright 2017-2019 Intel Corporation.
#
# This software and the related documents are Intel copyrighted  materials,  and
# your use of  them is  governed by the  express license  under which  they were
# provided to you (License).  Unless the License provides otherwise, you may not
# use, modify, copy, publish, distribute,  disclose or transmit this software or
# the related documents without Intel's prior written permission.
#
# This software and the related documents  are provided as  is,  with no express
# or implied  warranties,  other  than those  that are  expressly stated  in the
# License.
#
# License:
# http://software.intel.com/en-us/articles/intel-sample-source-code-license-agr
# eement/
#===============================================================================

#
#  Content:
#      Python sample of K-Means clustering in the distributed processing mode
#
#
from __future__ import print_function
import os
import sys

from pyspark import SparkContext, SparkConf

from daal import step1Local, step2Master
from daal.algorithms import kmeans
from daal.algorithms.kmeans import init

from distributed_hdfs_dataset import serializeNumericTable, DistributedHDFSDataSet, deserializePartialResult, deserializeNumericTable

utils_folder = os.path.join(os.environ['DAALROOT'], 'examples', 'python', 'source')
if utils_folder not in sys.path:
    sys.path.insert(0, utils_folder)
from utils import printNumericTable


nBlocks = 1
nClusters = 20
nIterations = 5
nVectorsInBlock = 10000


def runKmeans(dataRDD):

    partsRDD = computeInitLocal(dataRDD)
    centroids = computeInitMaster(partsRDD)

    for it in range(nIterations):
        serialized_centroids = serializeNumericTable(centroids)
        partsRDDcompute = computeLocal(dataRDD, serialized_centroids)
        centroids = computeMaster(partsRDDcompute)

    return centroids


def computeInitLocal(dataRDD):

    def mapper(tup):
        key, val = tup
        # Create an algorithm to initialize the K-Means algorithm on local nodes
        kmeansLocalInit = init.Distributed(step1Local,
                                           nClusters,
                                           nBlocks * nVectorsInBlock,
                                           nVectorsInBlock * key,
                                           method=init.randomDense)

        # Set the input data on local nodes
        deserialized_val = deserializeNumericTable(val)
        kmeansLocalInit.input.set(init.data, deserialized_val)

        # Initialize the K-Means algorithm on local nodes
        pres = kmeansLocalInit.compute()
        serialized_pres = serializeNumericTable(pres)

        return (key, serialized_pres)
    return dataRDD.map(mapper)


def computeInitMaster(partsRDD):

    # Create an algorithm to compute k-means on the master node
    kmeansMasterInit = init.Distributed(step2Master, nClusters, method=init.randomDense)

    partsList = partsRDD.collect()

    # Add partial results computed on local nodes to the algorithm on the master node
    for _, value in partsList:
        deserialized_pres = deserializePartialResult(value, init)
        kmeansMasterInit.input.add(init.partialResults, deserialized_pres)

    # Compute k-means on the master node
    kmeansMasterInit.compute()

    # Finalize computations and retrieve the results
    initResult = kmeansMasterInit.finalizeCompute()

    return initResult.get(init.centroids)


def computeLocal(dataRDD, centroids):

    def mapper(tup):

        key, val = tup

        # Create an algorithm to compute k-means on local nodes
        kmeansLocal = kmeans.Distributed(step1Local, nClusters, method=kmeans.defaultDense)

        # Set the input data on local nodes
        deserialized_val = deserializeNumericTable(val)
        deserialized_centroids = deserializeNumericTable(centroids)
        kmeansLocal.input.set(kmeans.data, deserialized_val)
        kmeansLocal.input.set(kmeans.inputCentroids, deserialized_centroids)

        # Compute k-means on local nodes
        pres = kmeansLocal.compute()
        serialized_pres = serializeNumericTable(pres)

        return (key, serialized_pres)
    return dataRDD.map(mapper)


def computeMaster(partsRDDcompute):

    # Create an algorithm to compute k-means on the master node
    kmeansMaster = kmeans.Distributed(step2Master, nClusters, method=kmeans.defaultDense)

    parts_List = partsRDDcompute.collect()

    # Add partial results computed on local nodes to the algorithm on the master node
    for _, value in parts_List:
        deserialized_pres = deserializePartialResult(value, kmeans)
        kmeansMaster.input.add(kmeans.partialResults, deserialized_pres)

    # Compute k-means on the master node
    kmeansMaster.compute()

    # Finalize computations and retrieve the results
    res = kmeansMaster.finalizeCompute()

    return res.get(kmeans.centroids)

if __name__ == "__main__":

    # Create SparkContext that loads defaults from the system properties and the classpath and sets the name
    sc = SparkContext(conf=SparkConf().setAppName("Spark Kmeans").setMaster('local[4]'))

    # Read from the distributed HDFS data set at a specified path
    dd = DistributedHDFSDataSet("/Spark/KmeansDense/data/")
    dataRDD = dd.getAsPairRDD(sc)

    # Compute k-means for dataRDD
    result = runKmeans(dataRDD)

    # Redirect stdout to a file for correctness verification
    stdout = sys.stdout
    sys.stdout = open('KmeansDense.out', 'w')

    # Print the results
    printNumericTable(result, "First 10 dimensions of centroids:", 20, 10)

    # Restore stdout
    sys.stdout = stdout

    sc.stop()
